### PR TITLE
Benchmarking, packaging, and fix issues, tests and perfomance

### PR DESCRIPTION
**Pull request by urbanishimwe**
### performance
- handling of the very big packet(any size that can be buffered)
- speeding up TCP sessions by using message hints: Added **proto.HasFullPayload** that helps to validate the entire HTTP request, it supports `Chunked` encoding too! Added **proto.HasRequestTitle** and **proto.HasResponseTitle** for validating the beginning of HTTP request. Those methods are used `input_raw.go` with `TCP`.
- supports Keep-Alive: the above functions helps to support keep-alive

### Packaging
- **capture:** engines(capture/doc.go)
- **tcp:** tcp message parser (tcp/doc.go)

### benchmarking
- **capture.BenchmarkPcapDump:** the benchmarks regarding dumping packets in a pcap file
- **capture.BenchmarkPcapFile:** the benchmarks of reading packets from a pcap file
- **capture.BenchmarkPcap:** the benchmarks of parsing packets from the loopback interface with pcap handles
- **proto.BenchmarkHasFullPayload:**: benchmarking this function which validates the HTTP payload
- **tcp.BenchmarkPacketParseAndSort:** benchmarks of parsing and sorting packets
- **tcp.BenchmarkMessageParserWithoutHint:** benchmarks of message reasembling by using `SYN` and `FIN` flag
- **tcp.BenchmarkMessageParserWithHint:** benchmarks of message reasembling by using `proto.HasRequestTitle` and `proto.HasFullPayload` flag

### issues
see linked issues

###  tests
- fixed input raw and engine tests

**Most of the changed of the files, was about using functionalities of** `tcp` **and** `capture` **in existing functionalities**

### CODE REVIEW

File: capture/capture_test.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+			return v
+		}
+	}
+	return ifis[0]
+}()
+
+func TestSetInterfaces(t *testing.T) {
+	l := &Listener{}
+	l.host = "127.0.0.1"
+	l.setInterfaces()
```
Comment (arijitAD) [inline capture/capture_test.go:29]: Check for errors. This is at multiple places.

File: capture/capture_test.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	}
+	name := f.Name()
+	f.Close()
+	var l *Listener
+	l, err = NewListener(name, 8000, "", EnginePcapFile, true)
+	if err != nil {
+		b.Error(err)
+		return
+	}
+	l.Activate()
```
Comment (arijitAD) [inline capture/capture_test.go:233]: Check for error.

File: emitter.go [hunk 3]
```diff
@@ -198,7 +191,7 @@ func CopyMulty(stop chan int, src io.Reader, writers ...io.Writer) error {
 					hasher := fnv.New32a()
 					// First 20 bytes contain tcp session
 					id := payloadID(payload)
-					hasher.Write(id[:20])
+					hasher.Write(id)
```
Comment (arijitAD) [inline emitter.go:194]: Check return value.

File: input_raw_test.go [hunk 4]
```diff
@@ -30,22 +27,29 @@ func TestRAWInputIPv4(t *testing.T) {
...
 	listener, err := net.Listen("tcp", ":0")
 	if err != nil {
-		t.Fatal(err)
+		t.Error(err)
+		return
 	}
 	origin := &http.Server{
-		Handler:      http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write([]byte("ab"))
```
Comment (arijitAD) [inline input_raw_test.go:35]: Error check. Multiple places.

File: capture/capture_test.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	}
+	<-waiter
+	name := f.Name()
+	f.Close()
+	testPcapDumpEngine(name, t)
+}
+
+func testPcapDumpEngine(f string, t *testing.T) {
+	defer os.Remove(f)
+	l, err := NewListener(f, 8000, "", EnginePcapFile, true)
```
Comment (arijitAD) [inline capture/capture_test.go:132]: Check if err != nil

File: capture/capture.go [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+	Interfaces []NetInterface
+	Reading    chan bool // this channel is closed when the listener has started reading packets
+
+	host          string // pcap file name or interface (name, hardware addr, index or ip address)
+	port          uint16 // src or/and dst port
+	trackResponse bool
+
+	quit      chan bool
+	packets   chan gopacket.Packet
+	activated bool
```
Comment (arijitAD) [inline capture/capture.go:56]: This is unused.

File: input_raw.go [hunk 7]
```diff
@@ -1,125 +1,209 @@
...
+type RAWInputConfig struct {
+	capture.PcapOptions
+	Expire         time.Duration      `json:"input-raw-expire"`
+	CopyBufferSize size.Size          `json:"copy-buffer-size"`
+	Engine         capture.EngineType `json:"input-raw-engine"`
+	TrackResponse  bool               `json:"input-raw-track-response"`
+	Protocol       TCPProtocol        `json:"input-raw-protocol"`
+	RealIPHeader   string             `json:"input-raw-realip-header"`
+	Stats          bool               `json:"input-raw-stats"`
+	quit           chan bool          // Channel used only to indicate goroutine should shutdown
```
Comment (arijitAD) [inline input_raw.go:62]: Unused

File: tcp/tcp_test.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+func TestMessageParserWithHint(t *testing.T) {
+	var mssg = make(chan *Message, 3)
+	pool := NewMessagePool(1<<20, time.Second, nil, func(m *Message) { mssg <- m })
+	pool.Start = func(pckt *Packet) (bool, bool) {
+		return proto.HasRequestTitle(pckt.Payload), proto.HasResponseTitle(pckt.Payload)
+	}
+	pool.End = func(m *Message) bool {
+		return proto.HasFullPayload(m.Data())
+	}
+	var packets []gopacket.Packet
```
Comment (arijitAD) [inline tcp/tcp_test.go:58]: Merge these two lines.

File: proto/proto_test.go [hunk 9]
```diff
@@ -332,4 +305,177 @@ func TestSetHostHTTP10(t *testing.T) {
...
+		payload[i] = '1'
+	}
+	var ok bool
+	data := []byte("HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nTransfer-Encoding: chunked\r\n\r\n")
+	if ok = HasFullPayload(data); ok {
+		b.Error("HasFullPayload should fail")
+		return
+	}
+	for i := 0; i < b.N; i++ {
+		data = append(data, append([]byte(fmt.Sprintf("fc00\r\n%s\r\n", payload)))...)
```
Comment (arijitAD) [inline proto/proto_test.go:469]: Internal `append` is not required.

File: capture/capture_test.go (hunk 5 above)
Comment (urbanishimwe) [reply capture/capture_test.go:132]: it's checked at the line 134

File: capture/capture_test.go (hunk 1 above)
Comment (urbanishimwe) [reply capture/capture_test.go:29]: we handled the check at line 30

File: input_raw_test.go (hunk 4 above)
Comment (urbanishimwe) [reply input_raw_test.go:35]: We won't really use it anywhere! it will unnecessary allocation

File: emitter.go (hunk 3 above)
Comment (urbanishimwe) [reply emitter.go:194]: We won't really use it anywhere! it will unnecessary allocation

File: input_raw.go (hunk 7 above)
Comment (urbanishimwe) [reply input_raw.go:62]: which one is unused?

**Pull request merged**